core/src/impl/Kokkos_Printf.hpp

```diff
@@ -0,0 +1,47 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_IMPL_PRINTF_HPP
+#define KOKKOS_IMPL_PRINTF_HPP
+
+#include <Kokkos_Macros.hpp>
+
+namespace Kokkos {
+
+// Considering an
+//
+// template <typename... Args>
+// KOKKOS_FUNCTION int printf(const char* format, Args... args);
+//
+// interface, Kokkos::printf for the Serial backend is implemented in terms of
+// std::printf printing the data specified in format and args... returning the
+// number of characters written if successful or a negative value if an error
+// encountered. The implementation for other backends emulates this behavior.
+#ifdef KOKKOS_ENABLE_SYCL
+using sycl::ext::oneapi::experimental::printf;
+#elif defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
+template <typename... Args>
+KOKKOS_FUNCTION int printf(const char* format, Args... args) {
+  ::printf(format, args...);
```

Comment:
Returning a value here doesn't work with `NVHPC`.

Reply:
Did we discuss whether we should be providing a checked version or return anything at all?
By that I mean, should we attempt to cause program termination when we detect than an error occurred?
Should we return void instead of "passing through" the return value of `printf` as currently proposed?

I don't necessarily have an issue with what is proposed but I do want to make sure we considered alternatives.

Reply:
I think it's good to stay close to what `printf` does since that's what people are used to (even though I doubt many people use the return value anyway).